cmd/kubeadm/app/cmd/phases/certs_test.go

```diff
@@ -0,0 +1,267 @@
+		}
+	}
+	for i, ip := range []string{"10.0.0.1", "1.2.3.4"} {
+		if APIserverCert.IPAddresses[i].String() != ip {
+			t.Errorf("APIserverCert.IPAddresses[%d] is %s instead of %s", i, APIserverCert.IPAddresses[i], ip)
+		}
+	}
+}
+
+func TestSubCmdReadsConfig(t *testing.T) {
+
+	subCmds := newSubCmdCerts()
+
+	var tests = []struct {
+		subCmds           []string
+		expectedFileCount int
+	}{
+		{
+			subCmds:           []string{"all"},
+			expectedFileCount: 12,
+		},
+		{
+			subCmds:           []string{"ca", "apiserver", "apiserver-kubelet-client"},
+			expectedFileCount: 6,
+		},
+		{
+			subCmds:           []string{"sa"},
+			expectedFileCount: 2,
+		},
+		{
+			subCmds:           []string{"front-proxy-ca", "front-proxy-client"},
+			expectedFileCount: 4,
+		},
+	}
+
+	for _, test := range tests {
+		// Temporary folder for the test case
+		tmpdir, err := ioutil.TempDir("", "")
+		if err != nil {
+			t.Fatalf("Couldn't create tmpdir")
+		}
+		defer os.RemoveAll(tmpdir)
+
+		configPath := saveDummyCfg(t, tmpdir)
+
+		// executes given sub commands
+		for _, subCmdName := range test.subCmds {
+			subCmd := getSubCmd(t, subCmdName, subCmds)
+			subCmd.SetArgs([]string{fmt.Sprintf("--config=%s", configPath)})
+			if err := subCmd.Execute(); err != nil {
+				t.Fatalf("Could not execute command: %s", subCmdName)
+			}
+		}
+
+		// verify expected files are there
+		assertFilesCount(t, tmpdir, test.expectedFileCount+1)
+	}
+}
+
+func getSubCmd(t *testing.T, name string, subCmds []*cobra.Command) *cobra.Command {
+	for _, subCmd := range subCmds {
+		if subCmd.Name() == name {
+			return subCmd
+		}
+	}
+	t.Fatalf("Unable to find sub command %s", name)
+
+	return nil
+}
+
+func assertFilesCount(t *testing.T, dirName string, count int) {
+	files, err := ioutil.ReadDir(dirName)
+	if err != nil {
+		t.Fatalf("Couldn't read files from tmpdir: %s", err)
+	}
+
+	if len(files) != count {
+		t.Errorf("dir does contains %d, %d expected", len(files), count)
+		for _, f := range files {
+			t.Error(f.Name())
+		}
+	}
+}
+
+func assertFileExists(t *testing.T, dirName string, fileName string) {
+	path := path.Join(dirName, fileName)
+	if _, err := os.Stat(path); os.IsNotExist(err) {
+		t.Errorf("file %s does not exist", fileName)
+	}
+}
+
+func saveDummyCfg(t *testing.T, dirName string) string {
+
+	path := path.Join(dirName, "dummyconfig.yaml")
+	cfgTemplate := template.Must(template.New("init").Parse(dedent.Dedent(`
+		apiVersion: kubeadm.k8s.io/v1alpha1
+		kind: MasterConfiguration
+		certificatesDir: {{.CertificatesDir}}
+		`)))
+
+	f, err := os.Create(path)
+	if err != nil {
+		t.Errorf("error creating dummyconfig file %s: %v", path, err)
+	}
+
+	templateData := struct {
+		CertificatesDir string
+	}{
+		CertificatesDir: dirName,
+	}
+
+	err = cfgTemplate.Execute(f, templateData)
+	if err != nil {
+		t.Errorf("error generating dummyconfig file %s: %v", path, err)
+	}
+	f.Close()
+
+	return path
+}
```

Comment:
+1, however I will take care of those implementation after addressing all the other remarks.